lib/Echidna/Solidity.hs

```diff
@@ -63,27 +64,34 @@ data SolConf = SolConf { _contractAddr :: Addr   -- ^ Contract address to use
                        , _sender       :: [Addr] -- ^ Sender addresses to use
                        , _prefix       :: Text   -- ^ Function name prefix used to denote tests
                        , _solcArgs     :: String -- ^ Args to pass to @solc@
+                       , _quiet        :: Bool   -- ^ Option to suppress some output and stderr of @solc@
                        }
 makeLenses ''SolConf
 
 -- | Given a file, try to compile it and get a list of its contracts, throwing exceptions if necessary.
 contracts :: (MonadIO m, MonadThrow m, MonadReader x m, Has SolConf x) => FilePath -> m [SolcContract]
-contracts fp = view (hasLens . solcArgs) >>= liftIO . solc >>= (\case
-  Nothing -> throwM CompileFailure
-  Just m  -> pure . toList $ fst m) where
-    solc a = readSolc =<< writeSystemTempFile "" =<< readProcess "solc" (usual <> words a) ""
-    usual = ["--combined-json=bin-runtime,bin,srcmap,srcmap-runtime,abi,ast", fp]
-    
+contracts fp = do
+  a <- view (hasLens . solcArgs)
+  stfu <- view (hasLens . quiet)
+  pure (a, stfu) >>= liftIO . solc >>= (\case
+    Nothing -> throwM CompileFailure
+    Just m  -> pure . toList $ fst m) where
+      solc (a, stfu) = do
+        stderr <- if stfu
+                  then UseHandle <$> openFile "/dev/null" WriteMode
+                  else pure Inherit
+        readSolc =<< writeSystemTempFile "" =<< readCreateProcess (proc "solc" (usual <> words a)) { std_err = stderr } ""
+      usual = ["--combined-json=bin-runtime,bin,srcmap,srcmap-runtime,abi,ast", fp]
+
 -- | Given a file and a possible contract name, compile the file as solidity, then, if a name is
 -- given, try to return the specified contract, otherwise, return the first contract in the file,
 -- throwing errors if necessary.
 selected :: (MonadIO m, MonadThrow m, MonadReader x m, Has SolConf x) => FilePath -> Maybe Text -> m SolcContract
 selected fp name = do cs <- contracts fp
                       c <- choose cs $ ((pack fp <> ":") <>) <$> name
-                      liftIO $ if isNothing name && length cs > 1
-                        then putStrLn "Multiple contracts found in file, only analyzing the first"
-                        else pure ()
-                      liftIO . putStrLn $ "Analyzing contract: " <> unpack (c ^. contractName)
+                      stfu <- view (hasLens . quiet)
+                      liftIO $ when (isNothing name && length cs > 1) $ putStrLn "Multiple contracts found in file, only analyzing the first"
```

Comment:
Let's get these two `when`s in one `liftIO`, and split lines like this to be a bit shorter. Also, general note, I think we should prefer `f . g $ h x` style (relevant a few places in this PR)

Reply:
Done.